chemprop/models/ffn.py

```diff
@@ -0,0 +1,439 @@
+            output_size=output_size,
+            dropout=dropout,
+            activation=activation,
+        )
+        self.weights_readout = build_ffn(
+            first_linear_dim=base_output_size,
+            hidden_size=hidden_size,
+            output_size=1,
+            num_layers=weights_ffn_num_layers,
+            dropout=dropout,
+            activation=activation,
+        )
+        self.ffn_type = ffn_type
+
+    def forward(
+        self,
+        input: Tuple[torch.Tensor, List, torch.Tensor, List, torch.Tensor],
+        constraints: torch.Tensor,
+        bond_types: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Runs the :class:`FFNAtten` on input.
+        :param input: A tuple of atom and bond informations of each molecule.
+        :param constraints: A PyTorch tensor which applies constraint on atomic/bond properties.
+        :param bond_types: A PyTorch tensor storing bond types of each bond determined by RDKit molecules.
+        :return: The output of the :class:`FFNAtten`, a PyTorch tensor containing a list of property predictions.
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == "atom":
+            hidden = a_hidden
+            scope = a_scope
+        elif self.ffn_type == "bond":
+            forward_bond = b_hidden[b2br[:, 0]]
+            backward_bond = b_hidden[b2br[:, 1]]
+            scope = [((start - 1) // 2, size // 2) for start, size in b_scope]
+        else:
+            raise RuntimeError(f"Unrecognized ffn_type of {self.ffn_type}.")
+
+        if self.ffn_type == "atom":
+            output_hidden = self.ffn(hidden)
+            output = self.ffn_readout(output_hidden)
+        elif self.ffn_type == "bond":
+            b_hidden_1 = torch.cat([forward_bond, backward_bond], dim=1)
+            b_hidden_2 = torch.cat([backward_bond, forward_bond], dim=1)
+            output_1 = self.ffn(b_hidden_1)
+            output_2 = self.ffn(b_hidden_2)
+            output_hidden = (output_1 + output_2) / 2
+            output = self.ffn_readout(output_hidden)
+            if bond_types is not None:
+                output = output + bond_types.reshape(-1, 1)
+
+        W_a = self.weights_readout(output_hidden)
+        constrained_output = []
+        for i, (start, size) in enumerate(scope):
+            if size == 0:
+                continue
+            else:
+                q_i = output[start:start+size]
+                w_i = W_a[start:start+size].softmax(0)
+                Q = constraints[i]
+                q_f = q_i + w_i * (Q - q_i.sum())
+                constrained_output.append(q_f)
+
+        output = torch.cat(constrained_output, dim=0)
+
+        return output
+
+
+class Exp(nn.Module):
+    def forward(self, x):
+        return x.exp()
+
+def build_ffn(
+    first_linear_dim: int,
+    hidden_size: int,
+    num_layers: int,
+    output_size: int,
+    dropout: float,
+    activation: nn.Module,
```

Comment:
can we use the `get_activation()` function defined in this package? Seems unnatural that some functions in the library would have the `activation` argument correspond to the _name_ of the activation function whereas for some functions it actually corresponds to the `Module` of the activation function

Reply:
That makes sense to me. I will apply this suggestion. Thanks a lot!